sdk/communication/Azure.Communication.Common/src/EntraTokenCredential.cs

```diff
@@ -0,0 +1,143 @@
+        /// <returns>
+        /// A task that represents the asynchronous get token operation. The value of its <see cref="ValueTask{AccessToken}.Result"/> property contains the access token.
+        /// </returns>
+        public ValueTask<AccessToken> GetTokenAsync(CancellationToken cancellationToken = default)
+            => _accessTokenCache.GetValueAsync(cancellationToken);
+
+        private AccessToken ExchangeEntraToken(CancellationToken cancellationToken)
+        {
+            return ExchangeEntraTokenAsync(false, cancellationToken).EnsureCompleted();
+        }
+
+        private async ValueTask<AccessToken> ExchangeEntraTokenAsync(CancellationToken cancellationToken)
+        {
+            var result = await ExchangeEntraTokenAsync(true, cancellationToken).ConfigureAwait(false);
+            return result;
+        }
+
+        private async ValueTask<AccessToken> ExchangeEntraTokenAsync(bool async, CancellationToken cancellationToken)
+        {
+            var message = CreateRequestMessage();
+            if (async)
+            {
+                await _pipeline.SendAsync(message, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                _pipeline.Send(message, cancellationToken);
+            }
+            return ParseAccessTokenFromResponse(message.Response);
+        }
+
+        private HttpMessage CreateRequestMessage()
+        {
+            var uri = new RequestUriBuilder();
+            uri.Reset(new Uri(_resourceEndpoint));
+            uri.AppendPath("/access/entra/:exchangeAccessToken", false);
+            uri.AppendQuery("api-version", "2024-04-01-preview", true);
+
+            var message = _pipeline.CreateMessage();
+            var request = message.Request;
+            request.Uri = uri;
+            request.Method = RequestMethod.Post;
+            request.Headers.Add("Accept", "application/json");
+            request.Headers.Add("Content-Type", "application/json");
+            request.Content = "{}";
+
+            return message;
+        }
+
+        private AccessToken ParseAccessTokenFromResponse(Response response)
+        {
+            switch (response.Status)
+            {
+                case 200:
+                    try
+                    {
+                        var json = JsonDocument.Parse(response.Content);
+                        var accessTokenJson = json.RootElement.GetProperty("accessToken").GetRawText();
+                        var acsToken = JsonSerializer.Deserialize<AcsToken>(accessTokenJson);
+                        return new AccessToken(acsToken.token, acsToken.expiresOn);
+                    }
+                    catch (Exception)
+                    {
+                        throw new RequestFailedException(response);
+                    }
+                default:
+                    throw new RequestFailedException(response);
+            }
+        }
+
+        private partial class AcsToken
```

Comment:
Why is this partial?

Reply:
I do not even remember why I put partial here😁